services/gem/gem.js

```diff
@@ -0,0 +1,322 @@
+
+
+class GemVersion extends BaseService {
+  async handle({repo}) {
+    const apiUrl = 'https://rubygems.org/api/v1/gems/' + repo + '.json';
+    const json = await this._sendAndCacheRequest(apiUrl, {
+      headers: { 'Accept': 'application/json' }
+    }).then(checkErrorResponse.asPromise())
+      .then(asJson);
+    const version = json.version;
+
+    return {
+      message: versionText(version),
+      color: versionColor(version)
+    };
+  }
+
+  // Metadata
+  static get defaultBadgeData() {
+    return { label: 'gem' };
+  }
+
+  static get category() {
+    return 'version';
+  }
+
+  static get url() {
+    return {
+      base: 'gem/v',
+      format: '(.*)',
+      capture: ['repo']
+    };
+  }
+
+  static get examples() {
+    return [
+      {
+        title: 'Gem',
+        previewUrl: 'formatador',
+        keywords: [
+          'ruby'
+        ]
+      }
+    ];
+  }
+};
+
+class GemDownloads extends BaseService {
+
+  _getApiUrl(repo, info) {
+    const endpoint = info === "dv" ? 'versions/' : 'gems/';
+    return `https://rubygems.org/api/v1/${endpoint}${repo}.json`;
+  }
+
+  _getLabel(version, info) {
+    if (version) {
+      return 'downloads@' + version;
+    } else {
+      if (info === "dtv") {
+        return 'downloads@latest';
+      } else {
+        return 'downloads';
+      }
+    }
+  }
+
+  async handle({info, rubygem}) {
+    const splitRubygem = rubygem.split('/');
+    const repo = splitRubygem[0];
+    let version = (splitRubygem.length > 1)
+      ? splitRubygem[splitRubygem.length - 1]
+      : null;
+    version = (version === "stable") ? version : semver.valid(version);
+    const label = this._getLabel(version, info);
+    const apiUrl = this._getApiUrl(repo, info);
+    const json = await this._sendAndCacheRequest(apiUrl, {
+      headers: { 'Accept': 'application/atom+json,application/json' }
+    }).then(checkErrorResponse.asPromise())
+      .then(asJson);
+
+    let downloads;
+    if (info === "dt") {
+      downloads = metric(json.downloads);
+    } else if (info === "dtv") {
+      downloads = metric(json.version_downloads);
+    } else if (info === "dv") {
+
+      let versionData;
+      if (version !== null && version === "stable") {
+
+        const versions = json.filter(function(ver) {
+          return ver.prerelease === false;
+        }).map(function(ver) {
+          return ver.number;
+        });
+        // Found latest stable version.
+        const stableVersion = latestVersion(versions);
+        versionData = json.filter(function(ver) {
+          return ver.number === stableVersion;
+        })[0];
+        downloads = metric(versionData.downloads_count);
+
+      } else if (version !== null) {
+
+        versionData = json.filter(function(ver) {
+          return ver.number === version;
+        })[0];
+
+        downloads = metric(versionData.downloads_count);
+      } else {
+        throw new InvalidResponse('invalid', new Error('version is null'));
```

Comment:
I'm passing a `new Error()` here because `InvalidResponse` expects `underlyingError` to exist. I suspect this is something we might end up doing a lot. Maybe it would be better if `underlyingError` were optional in the exception classes? Also `InvalidResponse` and `Inaccessible` [expect their params in different orders](https://github.com/badges/shields/blob/master/services/errors.js#L14-L27). I can see that there's a case for both params being optional. Maybe it would be better to change the signatures so we can pass both of them an object with optional keys `prettyMessage` and `underlyingError`? That would make it easier for either/both to be optional and make the interface more consistent.